File: androidApp/src/main/java/com/luisma/conexiones/android/features/levels/components/LevelsGrid.kt
package com.luisma.conexiones.android.features.levels.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.rememberLazyGridState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.luisma.conexiones.android.R
import com.luisma.conexiones.android.core_ui.components.CButton
import com.luisma.conexiones.android.core_ui.components.LevelCard
import com.luisma.conexiones.android.core_ui.components.LevelCardData
import com.luisma.conexiones.android.core_ui.components.LevelFloatingButton
import com.luisma.conexiones.android.core_ui.components.LevelLoadingCard
import com.luisma.conexiones.android.core_ui.components.levelsCardSize
import com.luisma.conexiones.android.core_ui.helpers.header
import com.luisma.conexiones.android.core_ui.theme.CThemeProvider
import com.luisma.conexiones.android.features.levels.LevelsEvents
import com.luisma.conexiones.android.features.levels.LevelsState
import com.luisma.conexiones.android.features.levels.mapLevelsState
import kotlinx.coroutines.launch


@Composable
fun LevelsGrid(
    state: LevelsState,
    sendEvent: (event: LevelsEvents) -> Unit
) {
    // ui
    val cardsPadding = 12.dp
    val gridPaddingBottom = 50.dp
    val levelsCardAmount = 2

    // loading cards
    val topLoadingHeader = if (state.reachFirstPage) 0 else 1
    val bottomLoadingCards =
        if (state.reachLastPage) 0 else levelsCardAmount

    // scroll
    val gridState = rememberLazyGridState()
    val coroutineScope = rememberCoroutineScope()

    LaunchedEffect(
        key1 = Unit
    ) {
        if (!state.initialScrollDone
            && state.initialScrollIdx > 0
        ) {
            gridState.scrollToItem(index = state.initialScrollIdx + topLoadingHeader)
            sendEvent(LevelsEvents.SetInitialScrollDone)
        }
    }

    // wrapper box: grid + floating button
    Box {
        // grid
        LazyVerticalGrid(
            columns = GridCells.FixedSize(
                size = levelsCardSize + cardsPadding * 2
            ),
            horizontalArrangement = Arrangement.Center,
            state = gridState,
            contentPadding = PaddingValues(bottom = gridPaddingBottom)
        ) {

            val cardModifier = Modifier.padding(
                all = cardsPadding,
            )

            // top loaders
            if (topLoadingHeader > 0) {
                header {
                    CButton(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(
                                horizontal = cardsPadding
                            ),
                        text = stringResource(id = R.string.levels_previous_levels_btn),
                        iconId = R.drawable.ic_arrow_up_solid,
                        onTap = {
                            sendEvent(LevelsEvents.GetPreviousPage)
                        }
                    )
                }
            }

            // grid levels
            items(
                count = state.games.count()
            ) { levelIdx ->
                val game = state.games[levelIdx]
                LevelCard(
                    modifier = cardModifier,
                    card = LevelCardData(
                        id = game.id,
                        type = mapLevelsState(game.gameState),
                        level = game.id.toString(),
                        livesReward = game.livesEarnedFormatted,
                    ),
                    checkVisibility = game.id == state.playingRowId,
                    onAppear = { sendEvent(LevelsEvents.OnVisibilityChangePlayingCard(show = true)) },
                    onDestroy = { sendEvent(LevelsEvents.OnVisibilityChangePlayingCard(show = false)) },
                    onTap = { id -> sendEvent(LevelsEvents.OnTapLevel(gameId = id)) }
                )
            }

            // bottom loaders
            if (bottomLoadingCards > 0) {
                items(
                    count = bottomLoadingCards
                ) { idx ->
                    LevelLoadingCard(
                        modifier = cardModifier,
                        idx = idx,
                        onAppear = { sendEvent(LevelsEvents.GetNextPage) },
                    )
                }
            }
        }

        // floating button
        if (!state.playingRowIsShowing) {
            LevelFloatingButton(
                modifier = Modifier.align(Alignment.BottomCenter),
                onTap = {
                    coroutineScope.launch {
                        gridState.scrollToItem(index = state.playingRowIdx)
                    }
                }
            )
        }
    }
}


@Preview
@Composable
private fun LevelsGridPreview() {
    CThemeProvider {
        LevelsGrid(
            state = LevelsState.initial(),
            sendEvent = {}
        )
    }
}

File: iosApp/iosApp/Features/Levels/Components/LevelsGrid.swift
//
//  LevelsGrid.swift
//  iosApp
//
//  Created by Luis Manuel García Fernández on 24/8/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import shared
import SwiftUI

let levelsSpacingGrid = 20.0
let levelsSizeGrid = levelCardSize + levelsSpacingGrid
let levelsSizePrevButton = 2 * levelsSizeGrid

struct LevelsGrid: View {
    @EnvironmentObject var vm: LevelsViewModel

    let columns = [
        GridItem(
            .adaptive(
                minimum: levelsSizeGrid,
                maximum: levelsSizeGrid
            )
        ),
        GridItem(
            .adaptive(
                minimum: levelsSizeGrid,
                maximum: levelsSizeGrid
            )
        ),
    ]

    func mapLevelCardType(state: GameState) -> LevelCardType {
        switch state {
        case GameState.win:
            return .Win
        case GameState.lost:
            return .Lost
        case GameState.free:
            return .Unlock
        case GameState.lock:
            return .Lock
        case GameState.playing:
            return .Playing
        default:
            return .Lock
        }
    }

    var body: some View {
        let bottomCardLoading = vm.state.reachLastPage ? 0 : 1
        let count = vm.state.reachLastPage ? vm.state.games.count : vm.state.games.count + 1
        return ScrollViewReader { scroll in
            ZStack(
                alignment: .bottom
            ) {
                ScrollView {
                    VStack {
                        // btn previous page
                        if !vm.state.reachFirstPage {
                            CButton(
                                width: levelsSizePrevButton,
                                svg: .arrowUp,
                                key: "levels_previous_levels_btn",
                                onTap: {
                                    vm.sendEvent(event: .previousPage)
                                }
                            )
                            .padding([.top, .bottom], cSpace8)
                        }

                        // grid levels
                        LazyVGrid(
                            columns: columns,
                            spacing: levelsSpacingGrid
                        ) {
                            ForEach(
                                0 ..< count,
                                id: \.self
                            ) { idx in
                                // btns loading and levels card
                                if bottomCardLoading == 1 && idx == vm.state.games.count {
                                    LevelCard(
                                        isLoading: true,
                                        cardData: LevelCardData(type: .Lock)
                                    )
                                    .onAppear {
                                        vm.sendEvent(event: .nextPage)
                                    }
                                } else {
                                    let game = vm.state.games[idx]
                                    LevelCard(
                                        cardData: LevelCardData(
                                            type: mapLevelCardType(state: game.gameState),
                                            level: String(game.id),
                                            livesReward: game.livesEarnedFormatted
                                        ),
                                        onTap: {
                                            vm.sendEvent(event: .goToLevel(gameId: Int(game.id)))
                                        }
                                    )
                                    .onAppear {
                                        if game.id == vm.state.playingRowId {
                                            vm.sendEvent(event: .onVisibilityChangePlayingCard(show: true))
                                        }
                                    }
                                    .onDisappear {
                                        if game.id == vm.state.playingRowId {
                                            vm.sendEvent(event: .onVisibilityChangePlayingCard(show: false))
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                .onAppear {
                    if !vm.state.initialScrollDone {
                        scroll.scrollTo(vm.state.initialScrollIdx, anchor: .top)
                        vm.sendEvent(event: .initialScrollDone)
                    }
                }
                // btn go to last played
                if !vm.state.playingRowIsShowing {
                    LevelFloatingButton(
                        onTap: { scroll.scrollTo(vm.state.playingRowIdx, anchor: .top) }
                    )
                }
            }
        }
    }
}
